Check ContainerID before Deleting a pod

gunjan5 opened the pull request:
Fixes https://github.com/projectcalico/cni-plugin/issues/310

gunjan5 commented:
not sure why Semaphore can't pull my libcalicogo fork, all tests pass locally (containerized)
```bash
Ran 28 of 28 Specs in 15.778 seconds
SUCCESS! -- 28 Passed | 0 Failed | 0 Pending | 0 Skipped PASS
```

tomdee commented on calico.go:
```diff
@@ -178,6 +178,7 @@ func cmdAdd(args *skel.CmdArgs) error {
 			endpoint = api.NewWorkloadEndpoint()
 			endpoint.Metadata.Name = args.IfName
 			endpoint.Metadata.Node = nodename
+			endpoint.Metadata.ActiveInstanceID = args.ContainerID
```
What's your reasoning for including this line?

gunjan5 replied:
we populate that field in `k8s.CmdAddK8s`, we want the same behavior for non-k8s orchestrators, this is for that.

tomdee replied:
Why do we want the same behaviour?

gunjan5 replied:
The reason why I am thinking this is what if some other orchestrator decides to be "aggressive" with DELs in future, we should be populating `ActiveInstanceID` similarly across all orchestrators so it's not confusing when someone looks at the code. Another reason is we go through different code paths for ADD (for k8s and non-k8s), but same code path for DEL and DEL code checks for `ActiveInstanceID` to avoid double-delete situation, which is unique to k8s for now, but could arise in future for any other orchestrators we support

tomdee replied:
But `ActiveInstanceID` is exactly the workload ID, which is already part of the key. It seems strange to store it twice.
The DEL code path should handle already it not being there.

tomdee commented on calico.go:
```diff
@@ -306,6 +307,33 @@ func cmdDel(args *skel.CmdArgs) error {
+		Workload:         workload,
+	}
+
+	wep, err := calicoClient.WorkloadEndpoints().Get(ep)
```
Does this `Get()` require the ActiveInstanceID to be present? What happens if it's missing (e.g. on upgrade?)

tomdee replied:
I was expecting more a of compare-and-delete approach here.

You at least need some comments explaining the complex flow we have going on here...

robbrockbank replied:
@tomdee @gunjan5 

Whose responsibility is it to police the ActiveInstanceID.  If its CNI then I'm worried there is a mighty big timing window here:

-  Old instance queries Endpoint gets ID - looks OK
-  New instance applies updated Endpoint data (updates the ActiveInstanceID)
-  Old instance goes - yup looks fine, delete.

At the moment, the consumer facing API does not include revision info so there is currently no CompareAndSwap.  That was something that I'd added in my libcalico-go refactor.  So there is no way for CNI to protect against this kind of operation.

Unless we can guarantee the delete is *always* called after a successful create/update then I worry this window will be quite large.

If you moved the ActiveInstanceID checking into libcalico-go then you do have the ability to do the CaS operation.

robbrockbank replied:
maybe nightly big is an exaggeration - but it might lead for some subtle hard to debug issues if the window is possible.

gunjan5 replied:
ActiveInstanceID is part of the Metadata, but not the key, so other 4 fields that make up the key are still unchanged

robbrockbank replied:
ActiveID is not required on the Get since it isn't an index.

tomdee commented on calico_cni_k8s_test.go:
```diff
@@ -602,6 +607,137 @@ var _ = Describe("CalicoCni", func() {
 				})
 			})
+
+			Context("ADD a continaer with a ContainerID and DEL it with the same ContainerID then ADD a new container with a different ContainerID, and send a DEL for the old ContainerID", func() {
```
typo

tomdee replied:
Add a comment with a link to the specific issues so we know why this test was added

tomdee commented on calico_cni_test.go:
```diff
@@ -52,16 +51,16 @@ var _ = Describe("CalicoCni", func() {
-			    "subnet": "10.0.0.0/8"
-			  }
-			}`, cniVersion, os.Getenv("ETCD_IP"))
+            {
```
Why change the indent here?

gunjan5 replied:
must have been my editor

tomdee replied:
So can you change it back?

tomdee commented on calico_cni_test.go:
```diff
@@ -297,4 +299,155 @@ var _ = Describe("CalicoCni", func() {
 		})
 	})
+
+	Describe("ADD a continaer with ContainerID and DEL it with the same ContainerID", func() {
```
typo

tomdee commented on calico_cni_test.go:
```diff
@@ -297,4 +299,155 @@ var _ = Describe("CalicoCni", func() {
+		})
+	})
+
+	Describe("ADD a continaer with a ContainerID and DEL it with the same ContainerID then ADD a new container with a different ContainerID, and send a DEL for the old ContainerID", func() {
```
typo

tomdee replied:
What value do you see in having a k8s and non-k8s version of this test?

gunjan5 replied:
right now it's just k8s being too "aggressive" in sending DELs, we could have other non-k8s doing the same and since both of these take different code paths (for ADD) I wanted to make sure we get the same behavior. In fact, I found an issue with this test, as the result I had L181 in calico.go :)

robbrockbank commented on calico.go:
```diff
@@ -339,16 +367,6 @@ func cmdDel(args *skel.CmdArgs) error {
-		Node:         nodename,
-		Orchestrator: orchestrator,
-		Workload:     workload}
 	if err = calicoClient.WorkloadEndpoints().Delete(ep); err != nil {
```
To do a atomic delete (in the sense no one else has modified it since you did your Get), you need to pass in the Metadata that was returned from the Get query - this will include revision data that etcd will use to determine whether the underlying entry has since been modified. 

So:  `calicoClient.WorkloadEndpoints().Delete(wep.Metadata)`

robbrockbank replied:
You probably want to explicitly handle an error type of `errors.ErrorResourceUpdateConflict` - I'm not sure if you need to retry the operation (I can't think that you'd need to), but you may want to log out that we're no-oping due to a concurrent update.

gunjan5 replied:
I don't think we want to retry, I don't see where/why we need to do that. I'm logging it and returning the error in that case.

tomdee commented on calico.go:
```diff
@@ -306,6 +307,32 @@ func cmdDel(args *skel.CmdArgs) error {
+	}
+
+	wep, err := calicoClient.WorkloadEndpoints().Get(ep)
+	if err != nil {
```
This is a behaviour change. Before, if a endpoint couldn't be found, we'd still try to clean up the IP address. Now we won't.

gunjan5 commented:
@tomdee I did some thinking about different cases and what clean up needs to be done for each case, my scribble below:
![image](https://cloud.githubusercontent.com/assets/5966968/25214619/f3f995c2-254c-11e7-88f9-2b889ecbf5ed.png)

I updated the code, and added a lot of comments around the DEL logic.

PTAL

tomdee commented on calico.go:
```diff
@@ -306,57 +306,84 @@ func cmdDel(args *skel.CmdArgs) error {
-		args.StdinData, err = json.Marshal(stdinData)
+	// The following logic only applies to kubernetes since it sends multiple DELs for the same endpoint.
+	// No need to get WEP from the datastore and compare it's ActiveInstanceID with ContainerID for non-k8s orchestrators
+	if orchestrator == "k8s" {
```
Can this block live in k8s.go since it's kubernetes specific?

gunjan5 replied:
we don't have a DEL func specific to k8s in k8s.go like we do for ADD, but I can move this part there if you'd like

tomdee replied:
Yes please

tomdee commented on calico.go:
```diff
@@ -306,57 +306,84 @@ func cmdDel(args *skel.CmdArgs) error {
 
-		args.StdinData, err = json.Marshal(stdinData)
+	// The following logic only applies to kubernetes since it sends multiple DELs for the same endpoint.
+	// No need to get WEP from the datastore and compare it's ActiveInstanceID with ContainerID for non-k8s orchestrators
```
...why? This comment would be more useful if you say why. e.g. ...because the ActiveInstanceID and ContainerID are the same value.

gunjan5 replied:
I've explained the "why" inside the code block in comments where I'm doing specific operations, but I'll add a line here too

tomdee commented on calico.go:
```diff
@@ -306,57 +306,84 @@ func cmdDel(args *skel.CmdArgs) error {
-		// It just needs a valid CIDR, but it doesn't have to be the CIDR associated with the host.
-		dummyPodCidr := "0.0.0.0/0"
-		var stdinData map[string]interface{}
+	calicoClient, err := CreateClient(conf)
```
Are you sure that creating a client doesn't actually connect to the datastore?

gunjan5 replied:
It just creates the client, doesn't actually connect to the datastore. This was there before my PR, so no change in behavior here

tomdee commented on calico.go:
```diff
@@ -306,57 +306,84 @@ func cmdDel(args *skel.CmdArgs) error {
+
+				// We can talk to the datastore but WEP doesn't exist in there,
+				// but we still want to go ahead with the clean up. So, log a warning and clean up.
+				logger.WithField("WorkloadEndpoint", ep).Warningf("Error getting workloadendpoint: %v", err)
```
A warning log, with "error" in the text? That's probably not a good idea. I think it would be good if the log also say that processing will continue.

tomdee commented on calico.go:
```diff
@@ -306,57 +306,84 @@ func cmdDel(args *skel.CmdArgs) error {
+		// Check if ActiveInstanceID is populated (it will be an empty string "" if it was populated
+		// before this field was added to the API), and if it is there then compare it with ContainerID
+		// passed by the orchestrator to make sure they are the same, return without deleting if they aren't.
+		if endpoint.ActiveInstanceID != "" && args.ContainerID != endpoint.ActiveInstanceID {
```
This code assumes that an ADD and DEL for the same workload ID (but different container ID) can't happen at the same time. If that is correct, that you should document that assumption in a comment.

tomdee replied:
I think this comment got lost

tomdee commented on calico.go:
```diff
@@ -306,57 +306,84 @@ func cmdDel(args *skel.CmdArgs) error {
-	calicoClient, err := CreateClient(conf)
-	if err != nil {
-		return err
+	// Store WEP Metadata in endpoint only if it is not nil to avoid nil pointer dereference
```
Are you sure you need this? I don't think you do...

gunjan5 replied:
yes, it's needed because in case of non-k8s the value of `endpoint` is an empty struct of type `api.WorkloadEndpoint` which won't work, so we need to use the initial `ep` we created to delete the WEP

tomdee replied:
There is something not right about this code. This block is duplicated with the section above, and the comment doesn't really explain why this is going on. Having `ep`, `wep` and `endpoint` doesn't help the situation.

tomdee commented on calico.go:
```diff
@@ -306,57 +306,84 @@ func cmdDel(args *skel.CmdArgs) error {
+			// Log and proceed with the clean up if WEP doesn't exist.
+			logger.WithField("endpoint", endpoint).Info("Endpoint object does not exist, no need to clean up.")
+		case errors.ErrorResourceUpdateConflict:
+			// This case means the WEP object was modified between the time we did the Get and now,
```
Why _don't_ you need to retry here?

gunjan5 replied:
I don't think retrying makes sense from CNI side since if some data changed while we were trying to Compare-and-Delete then retrying with the same data will still fail. 
It is the orchestrator's (k8s) job to send another DEL with the correct/different ContainerID if CNI plugin returns an error when k8s sent the DEL. k8s will do the retry here

tomdee replied:
The important point is that we don't care about all the data changing - we only really care about the containerID/activeworkloadID. Are you sure that k8s will retry and we won't leak endpoints? This is related to my comment elsewhere about concurrent ADD/DEL requests

tomdee commented on calico.go:
```diff
@@ -306,57 +306,84 @@ func cmdDel(args *skel.CmdArgs) error {
+	// Release the IP address by calling the configured IPAM plugin.
+	ipamErr := cleanUpIPAM(conf, args, logger)
+
 	// Only try to delete the device if a namespace was passed in.
```
Since you've pulled cleanUpIpam into its own function, does it make sense to put the namespace cleanup in its own function too?

tomdee commented on calico_cni_k8s_test.go:
```diff
@@ -602,6 +619,147 @@ var _ = Describe("CalicoCni", func() {
 				})
 			})
+
+			// This specific test case is for an issue where k8s would send extra DELs being "aggrassive". See: https://github.com/kubernetes/kubernetes/issues/44100
```
typo - aggrassive

tomdee commented on calico.go:
```diff
@@ -306,57 +306,64 @@ func cmdDel(args *skel.CmdArgs) error {
-		// It just needs a valid CIDR, but it doesn't have to be the CIDR associated with the host.
-		dummyPodCidr := "0.0.0.0/0"
-		var stdinData map[string]interface{}
+	endpoint := &ep
```
Why do you have an endpoint and an `ep`?

gunjan5 replied:
endpoint's value is ep for non-k8s orch and for k8s when we can't get WEP, so the Delete operation doesn't error out, and can still clean up. endpoint's value is wep.Metadata so if we got WEP back from the datastore then we want to use it to have the Revision to safely do a Compare-and-delete

tomdee replied:
This isn't very clear from the naming of the variables. You need to name the variables better and/or comment it better.

tomdee commented on k8s/k8s.go:
```diff
@@ -294,6 +296,45 @@ func CmdAddK8s(args *skel.CmdArgs, conf utils.NetConf, nodename string, calicoCl
+// We store CNI_CONTAINERID as ActiveInstanceID in WEP Metadata for k8s,
+// so we need to make sure we check if ContainerID and ActiveInstanceID are the same before deleting the pod.
+func CmdDelK8s(c *calicoclient.Client, ep api.WorkloadEndpointMetadata, args *skel.CmdArgs, logger *log.Entry) (*api.WorkloadEndpointMetadata, error) {
+
```
I think we've had this before, but I don't like like the first line of a function to be blank.

tomdee commented on k8s/k8s.go:
```diff
@@ -294,6 +296,45 @@ func CmdAddK8s(args *skel.CmdArgs, conf utils.NetConf, nodename string, calicoCl
+	}
+
+	// Only compare ContainerID with ActiveInstanceID if we got WEP from the datastore.
+	if wep != nil {
```
Would this be better as an `else` to the block above?

gunjan5 replied:
both cases in the `if` above that `return` so there is no point of adding an else, it just makes the code harder to read. Also the linter yells :)

tomdee replied:
No they don't both return, which is why you need the extra if check.

robbrockbank replied:
If you added a `return nil, nil` to the ErrorResourceDoesNotExist branch then you could lose the `web != nil` check and unindent this whole section.

tomdee commented on k8s/k8s.go:
```diff
@@ -294,6 +296,45 @@ func CmdAddK8s(args *skel.CmdArgs, conf utils.NetConf, nodename string, calicoCl
+// The following logic only applies to kubernetes since it sends multiple DELs for the same endpoint.
+// We store CNI_CONTAINERID as ActiveInstanceID in WEP Metadata for k8s,
+// so we need to make sure we check if ContainerID and ActiveInstanceID are the same before deleting the pod.
+func CmdDelK8s(c *calicoclient.Client, ep api.WorkloadEndpointMetadata, args *skel.CmdArgs, logger *log.Entry) (*api.WorkloadEndpointMetadata, error) {
```
Does this name describe what the function is doing? To me it seems that it's not deleting anything, it's just fetching a WEPMeta?

tomdee commented on calico.go:
```diff
@@ -306,57 +306,68 @@ func cmdDel(args *skel.CmdArgs) error {
 
-		logger.WithField("podCidr", dummyPodCidr).Info("Using a dummy podCidr to release the IP")
-		stdinData["ipam"].(map[string]interface{})["subnet"] = dummyPodCidr
+		if wep != nil {
```
I think it's slightly abnormal for a function to return a nil error *and* and nil value, so probably worth a comment to note that.

tomdee commented on glide.yaml:
```diff
@@ -15,7 +15,7 @@ import:
   - gexec
 - package: github.com/projectcalico/libcalico-go
-  version: v1.1.0
+  version: 0a0ff81244b014c0d4076573edcc9e73694e673a
```
Isn't there a release with the fixes in?

gunjan5 replied:
it's an RC not an actual release yet

robbrockbank replied:
Yup:  v1.2.1-rc4

robbrockbank commented on calico.go:
```diff
@@ -306,57 +306,68 @@ func cmdDel(args *skel.CmdArgs) error {
+		if err != nil {
+			if err == k8s.ContainerIDMismatchErr {
+				// CNI_ContainerID does not match WorkloadEndpoint ActiveInstanceID so ignoring the DELETE cmd.
+				return nil
```
When k8s recreates the container - does it use the same IP address, or should we still be going through the IPAM deletion processing.

gunjan5 replied:
k8s doesn't recreate with the same IP, it does the ADD as if it's a new pod, so we still need to do the IPAM clean up

robbrockbank commented on calico.go:
```diff
@@ -306,57 +306,68 @@ func cmdDel(args *skel.CmdArgs) error {
-		args.StdinData, err = json.Marshal(stdinData)
-		if err != nil {
+	// Delete the WorkloadEndpoint object from the datastore.
+	if err = calicoClient.WorkloadEndpoints().Delete(*endpoint); err != nil {
```
Can we just move the actual deletion of the endpoint into the k8s code too - so basically once you call CmdDelK8s it will actually have deleted the WEP.  I think that would be more symmetric with the ADD no?

Yes it means you have two bits of code to handle the delete, one in k8s and one here for non-k8s - but that means wouldn't need to worry about resource update conflicts here (just in the k8s code where it makes more sense).

robbrockbank replied:
Also, although I can't think of an actual scenario where it is likely to happen, in the k8s code when you do have the resource update conflict I think you should retry as Tom suggests.  You need to retry from the Get() so that the revision and active instance ID are up to date.  If the WEP has had the active instance ID updated then you'll know about it and be able to log that with certainty.

gunjan5 replied:
Retrying in CNI doesn't make sense for a couple of reasons:
1. calling CNI to setUp/tearDown pod is like a one time job, so if it fails kubelet retries with the latest CNI ADD/DEL parameters, so CNI keep retrying with what info we have (which might change from k8s Pov) defeats that purpose
2. This would be a change in behavior, CNI returns an error if it can't delete the wep instead of retrying (i know it's under different case this time since we do a get, and CaD, but it's still a change in behavior)

I'm mostly saying no because of the first reason I stated above

robbrockbank commented on k8s/k8s.go:
```diff
@@ -294,6 +296,45 @@ func CmdAddK8s(args *skel.CmdArgs, conf utils.NetConf, nodename string, calicoCl
+		if _, ok := err.(cerrors.ErrorResourceDoesNotExist); ok {
+			// We can talk to the datastore but WEP doesn't exist in there,
+			// but we still want to go ahead with the clean up. So, log a warning and clean up.
+			logger.WithField("WorkloadEndpoint", ep).Warning("WorkloadEndpoint does not exist in the datastore, moving forward with the clean up")
```
I _think_ you mean to return nil, nil here (hence your comment below).  If that is the case then:

You can lose the next `else` and you can drop the `if wep != nil {` below and unindent that whole section.

gunjan5 replied:
yeah that makes sense

robbrockbank commented on k8s/k8s.go:
```diff
@@ -294,6 +296,45 @@ func CmdAddK8s(args *skel.CmdArgs, conf utils.NetConf, nodename string, calicoCl
 }
 
+// ContainerIDMismatchErr is for when ContainerID does not match ActiveInstanceID during Deletion process.
+var ContainerIDMismatchErr = errors.New("ContainerID does not match ActiveInstanceID")
```
I think the concept of the Container ID mismatch should be confined to the k8s code.  If you performed the remainder of the k8s WEP delete processing in CmdDelK8s you could achieve that.

tomdee commented on calico.go:
```diff
@@ -306,78 +301,40 @@ func cmdDel(args *skel.CmdArgs) error {
+	if orchestrator == "k8s" {
+		return k8s.CmdDelK8s(calicoClient, wep, args, conf, logger)
+	}
+
```
Previously, the IPAM delete happened _before_ the WEP delete. Why have you moved it?

gunjan5 replied:
that is because if the deleting WEP fails we return an error so DEL operation doesn't go through, I thought moving IPAM deletion after makes sense, but I guess it makes more sense in k8s.go not so much here

tomdee commented on calico_cni_test.go:
```diff
@@ -297,4 +308,61 @@ var _ = Describe("CalicoCni", func() {
 		})
 	})
+
+	Describe("ADD a container with ContainerID and DEL it with the same ContainerID", func() {
```
How is this test different from the first test in this file? Do we really need both?

tomdee commented on k8s/k8s.go:
```diff
@@ -294,6 +296,69 @@ func CmdAddK8s(args *skel.CmdArgs, conf utils.NetConf, nodename string, calicoCl
+	// In case of k8s, where we are deleting the WEP we got from the Datastore,
+	// this Delete is a Compare-and-Delete, so if *any* field in the WEP changed from
+	// the time we get WEP until here then the Delete operation will fail.
+	if err = c.WorkloadEndpoints().Delete(wep.Metadata); err != nil {
```
Why do we try to delete even if the get just failed?

gunjan5 replied:
I added an else there so it only tries to delete when get succeeds (forgot to save and commit it)

tomdee commented on k8s/k8s.go:
```diff
@@ -294,6 +296,69 @@ func CmdAddK8s(args *skel.CmdArgs, conf utils.NetConf, nodename string, calicoCl
+	// Return the IPAM error if there was one. The IPAM error will be lost if there was also an error in cleaning up
+	// the device or endpoint, but crucially, the user will know the overall operation failed.
+	return ipamErr
+
```
remove this blank line

tomdee commented on utils/utils.go:
```diff
@@ -40,6 +44,75 @@ func min(a, b int) int {
+
+	// We need to replace "usePodCidr" with a valid, but dummy podCidr string with "host-local" IPAM.
+	if conf.IPAM.Type == "host-local" && strings.EqualFold(conf.IPAM.Subnet, "usePodCidr") {
+
```
remove this blank line

tomdee commented on utils/utils.go:
```diff
@@ -40,6 +44,75 @@ func min(a, b int) int {
+func CleanUpIPAM(conf NetConf, args *skel.CmdArgs, logger *log.Entry) error {
+	var err error
+
+	// Always try to release the address. Don't deal with any errors till the endpoints are cleaned up.
```
This comment feels out of place

tomdee commented on utils/utils.go:
```diff
@@ -40,6 +44,75 @@ func min(a, b int) int {
+// CleanUpIPAM calls IPAM plugin to release the IP address.
+// It also contains IPAM plugin specific changes needed before calling the plugin.
+func CleanUpIPAM(conf NetConf, args *skel.CmdArgs, logger *log.Entry) error {
+	var err error
```
Why is this up here?

tomdee commented on k8s/k8s.go:
```diff
@@ -322,13 +322,12 @@ func CmdDelK8s(c *calicoclient.Client, ep api.WorkloadEndpointMetadata, args *sk
-	// this Delete is a Compare-and-Delete, so if *any* field in the WEP changed from
-	// the time we get WEP until here then the Delete operation will fail.
-	if err = c.WorkloadEndpoints().Delete(wep.Metadata); err != nil {
+		// Delete the WorkloadEndpoint object from the datastore.
```
These comments are indented too much

tomdee replied:
...so are the ones on 319:321

gunjan5 replied:
that's all `go fmt`

tomdee replied:
👍

tomdee commented on utils/utils.go:
```diff
@@ -75,22 +75,19 @@ func CleanUpNamespace(args *skel.CmdArgs, logger *log.Entry) error {
 		var stdinData map[string]interface{}
 
-		if err := json.Unmarshal(args.StdinData, &stdinData); err != nil {
+		err := json.Unmarshal(args.StdinData, &stdinData)
```
nit: I don't really understand why you changed these lines but I don't think it matters too much

gunjan5 replied:
this is because of line 97 where args.StdinData: `non-name args.StdinData on left side of :=
` so have to reuse the `err` variable.

tomdee replied:
OK, makes sense. I didn't see that bit of code.
